File: Unreal/Source/Network/Private/Utils/LZ4.cpp
/*
 * LZ4 Compression
 *
 * Author: Andre Ferreira
 *
 * Copyright (c) Uzmi Games. Licensed under the MIT License.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

 #include "Utils/LZ4.h"
 #include <cstring>
 
 int32 FLZ4::Compress(const uint8* Src, int32 SrcSize, uint8* Dst, int32 DstCapacity)
 {
     uint32 HashTable[HASH_SIZE] = {0};
     const uint8* Ip = Src;
     const uint8* Anchor = Ip;
     const uint8* Iend = Src + SrcSize;
     const uint8* Mflimit = Iend - MINMATCH;
     uint8* Op = Dst;
     uint8* Oend = Dst + DstCapacity;
 
     auto Hash = [](uint32 V) { return (V * 2654435761u) >> ((MINMATCH * 8) - HASH_LOG); };
 
     while (Ip < Mflimit)
     {
         uint32 Sequence = *(const uint32*)Ip;
         int H = Hash(Sequence);
         const uint8* Ref = Src + HashTable[H];
         HashTable[H] = (uint32)(Ip - Src);
 
         if (Ref < Ip && (Ip - Ref) < 0xFFFF && *(const uint32*)Ref == Sequence)
         {
             uint8* Token = Op++;
             int LiteralLength = (int)(Ip - Anchor);
 
             if (Op + LiteralLength + 8 > Oend)
                 return 0;
 
             if (LiteralLength >= 15)
             {
                 *Token = 15 << 4;
                 int Len = LiteralLength - 15;
 
                 while (Len >= 255)
                 {
                     *Op++ = 255;
                     Len -= 255;
                 }
 
                 *Op++ = (uint8)Len;
             }
             else
             {
                 *Token = (uint8)(LiteralLength << 4);
             }
 
             FMemory::Memcpy(Op, Anchor, LiteralLength);
             Op += LiteralLength;
 
             uint16 Offset = (uint16)(Ip - Ref);
             *Op++ = (uint8)Offset;
             *Op++ = (uint8)(Offset >> 8);
 
             Ip += MINMATCH;
             Ref += MINMATCH;
 
             int MatchLength = 0;
 
             while (Ip < Mflimit && *(const uint32*)Ip == *(const uint32*)Ref)
             {
                 Ip += 4;
                 Ref += 4;
                 MatchLength += 4;
             }
 
             while (Ip < Iend && *Ip == *Ref)
             {
                 ++Ip;
                 ++Ref;
                 ++MatchLength;
             }
 
             if (MatchLength >= 15)
             {
                 *Token |= 15;
                 int Len = MatchLength - 15;
                 while (Len >= 255)
                 {
                     *Op++ = 255;
                     Len -= 255;
                 }
                 *Op++ = (uint8)Len;
             }
             else
             {
                 *Token |= (uint8)MatchLength;
             }
 
             Anchor = Ip;
 
             if (Op > Oend - 5)
                 return 0;
 
             continue;
         }
         ++Ip;
     }
 
     int LastLit = (int)(Iend - Anchor);
 
     if (Op + LastLit + (LastLit / 255) + 1 > Oend)
         return 0;
 
     if (LastLit >= 15)
     {
         *Op++ = 15 << 4;
         int Len = LastLit - 15;
 
         while (Len >= 255)
         {
             *Op++ = 255;
             Len -= 255;
         }
 
         *Op++ = (uint8)Len;
     }
     else
     {
         *Op++ = (uint8)(LastLit << 4);
     }
 
     FMemory::Memcpy(Op, Anchor, LastLit);
     Op += LastLit;
 
     return int32(Op - Dst);
 }
 
 int32 FLZ4::Decompress(const uint8* Src, int32 SrcSize, uint8* Dst, int32 DstCapacity)
 {
     const uint8* Ip = Src;
     const uint8* Iend = Src + SrcSize;
     uint8* Op = Dst;
     uint8* Oend = Dst + DstCapacity;
 
     while (Ip < Iend)
     {
         uint8 Token = *Ip++;
         int LiteralLength = Token >> 4;
 
         if (LiteralLength == 15)
         {
             uint8 S;
             while (((S = *Ip++) == 255) && Ip < Iend)
                 LiteralLength += 255;
             LiteralLength += S;
         }
 
         if (Op + LiteralLength > Oend || Ip + LiteralLength > Iend)
             return 0;
 
         FMemory::Memcpy(Op, Ip, LiteralLength);
         Ip += LiteralLength;
         Op += LiteralLength;
 
         if (Ip >= Iend)
             break;
 
         uint16 Offset = (uint16)(Ip[0] | (Ip[1] << 8));
         Ip += 2;
         uint8* Match = Op - Offset;
 
         int MatchLength = Token & 0x0F;
 
         if (MatchLength == 15)
         {
             uint8 S;
             while (((S = *Ip++) == 255) && Ip < Iend)
                 MatchLength += 255;
             MatchLength += S;
         }
 
         MatchLength += 4;
 
         if (Op + MatchLength > Oend)
             return 0;
 
         while (MatchLength--)
         {
             *Op++ = *Match++;
         }
     }
 
     return int32(Op - Dst);
 }
 

File: Core/Utils/LZ4.cs
/*
 * LZ4 Compression
 *
 * Author: Andre Ferreira
 *
 * Copyright (c) Uzmi Games. Licensed under the MIT License.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

using System.Runtime.CompilerServices;

namespace Wormhole
{
    public static class LZ4
    {
        private const int MinMatch = 4;
        private const int HashLog = 16;
        private const int HashSize = 1 << HashLog;
        private const uint HashMultiplier = 2654435761u;

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static int Hash(uint value)
        {
            return (int)((value * HashMultiplier) >> ((MinMatch * 8) - HashLog));
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static unsafe int Compress(byte* src, int srcLength, byte* dst, int dstCapacity)
        {
            uint* hashTable = stackalloc uint[HashSize];

            for (int i = 0; i < HashSize; i++)
                hashTable[i] = 0;

            byte* ip = src;
            byte* anchor = ip;
            byte* iend = src + srcLength;
            byte* mflimit = iend - MinMatch;
            byte* op = dst;
            byte* oend = dst + dstCapacity;

            while (ip < mflimit)
            {
                uint sequence = *(uint*)ip;
                int h = Hash(sequence);
                byte* refp = src + hashTable[h];
                hashTable[h] = (uint)(ip - src);

                if (refp < ip && (ip - refp) < 0xFFFF && *(uint*)refp == sequence)
                {
                    byte* token = op++;
                    int literalLength = (int)(ip - anchor);

                    if (op + literalLength + 8 > oend)
                        return 0;

                    if (literalLength >= 15)
                    {
                        *token = 15 << 4;
                        int len = literalLength - 15;

                        while (len >= 255)
                        {
                            *op++ = 255;
                            len -= 255;
                        }

                        *op++ = (byte)len;
                    }
                    else
                    {
                        *token = (byte)(literalLength << 4);
                    }

                    for (int i = 0; i < literalLength; i++)
                        *op++ = anchor[i];

                    ushort offset = (ushort)(ip - refp);
                    *op++ = (byte)offset;
                    *op++ = (byte)(offset >> 8);

                    ip += MinMatch;
                    refp += MinMatch;

                    int matchLength = 0;

                    while (ip < mflimit && *(uint*)ip == *(uint*)refp)
                    {
                        ip += 4;
                        refp += 4;
                        matchLength += 4;
                    }

                    while (ip < iend && *ip == *refp)
                    {
                        ip++;
                        refp++;
                        matchLength++;
                    }

                    if (matchLength >= 15)
                    {
                        *token |= 15;
                        int len = matchLength - 15;

                        while (len >= 255)
                        {
                            *op++ = 255;
                            len -= 255;
                        }

                        *op++ = (byte)len;
                    }
                    else
                    {
                        *token |= (byte)matchLength;
                    }

                    anchor = ip;

                    if (op > oend - 5)
                        return 0;

                    continue;
                }

                ip++;
            }

            int lastLit = (int)(iend - anchor);

            if (op + lastLit + (lastLit / 255) + 1 > oend)
                return 0;

            if (lastLit >= 15)
            {
                *op++ = 15 << 4;
                int len = lastLit - 15;

                while (len >= 255)
                {
                    *op++ = 255;
                    len -= 255;
                }

                *op++ = (byte)len;
            }
            else
            {
                *op++ = (byte)(lastLit << 4);
            }

            for (int i = 0; i < lastLit; i++)
                *op++ = anchor[i];

            return (int)(op - dst);
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static unsafe int Decompress(byte* src, int srcLength, byte* dst, int dstCapacity)
        {
            byte* ip = src;
            byte* iend = src + srcLength;
            byte* op = dst;
            byte* oend = dst + dstCapacity;

            while (ip < iend)
            {
                byte token = *ip++;
                int literalLength = token >> 4;

                if (literalLength == 15)
                {
                    byte s;

                    while (((s = *ip++) == 255) && ip < iend)
                        literalLength += 255;

                    literalLength += s;
                }

                if (op + literalLength > oend || ip + literalLength > iend)
                    return 0;

                for (int i = 0; i < literalLength; i++)
                    *op++ = *ip++;

                if (ip >= iend)
                    break;

                ushort offset = (ushort)(ip[0] | (ip[1] << 8));
                ip += 2;
                byte* match = op - offset;
                int matchLength = token & 0x0F;

                if (matchLength == 15)
                {
                    byte s;

                    while (((s = *ip++) == 255) && ip < iend)
                        matchLength += 255;

                    matchLength += s;
                }

                matchLength += 4;

                if (op + matchLength > oend)
                    return 0;

                for (int i = 0; i < matchLength; i++)
                    *op++ = match[i];
            }

            return (int)(op - dst);
        }
    }
}